File: spec/test_basic_math_grammar.c
#include "testparser.h"
#include "testutils.h"
#include <string.h>

int main()
{
    char const * input = "1 + 2 * 3 + 4";
    p_context_t * context;
    context = p_context_new((uint8_t const *)input, strlen(input));
    assert_eq(P_SUCCESS, p_parse(context));
    assert_eq(11, p_result(context));
    p_context_delete(context);

    input = "1 * 2 ** 4 * 3";
    context = p_context_new((uint8_t const *)input, strlen(input));
    assert_eq(P_SUCCESS, p_parse(context));
    assert_eq(48, p_result(context));
    p_context_delete(context);

    input = "(1 + 2) * 3 + 4";
    context = p_context_new((uint8_t const *)input, strlen(input));
    assert_eq(P_SUCCESS, p_parse(context));
    assert_eq(13, p_result(context));
    p_context_delete(context);

    input = "(2 * 2) ** 3 + 4 + 5";
    context = p_context_new((uint8_t const *)input, strlen(input));
    assert_eq(P_SUCCESS, p_parse(context));
    assert_eq(73, p_result(context));
    p_context_delete(context);

    return 0;
}

File: spec/test_basic_math_grammar.rs
use testparser::*;

fn main() {
    let cases: [(&[u8], u64); 4] = [
        (b"1 + 2 * 3 + 4", 11),
        (b"1 * 2 ** 4 * 3", 48),
        (b"(1 + 2) * 3 + 4", 13),
        (b"(2 * 2) ** 3 + 4 + 5", 73),
    ];
    for (input, expected) in cases {
        let mut context = p_context_new(input);
        assert_eq!(P_SUCCESS, p_parse(&mut context));
        assert_eq!(expected, p_result(&context));
        p_context_delete(context);
    }
}
